src/hooks/useStepsContext.ts

```diff
@@ -88,6 +96,108 @@ export function useStepsContext(): IStepsContext {
         return oldSteps;
       });
     },
+    /**
+     * Used in conjunction with the drag/drop functionality for
+     * moving a step separator in the UI to reorganize the bucketing
+     * of actions.
+     *
+     * @example
+     * The examples below depend on this array:
+     *
+     * [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
+     *
+     *
+     * Example 1:
+     * Target index: 0
+     * Initiator index: 1
+     * Action index: 1
+     * Output: [[1, 2], [3, 4, 5, 6], [7, 8, 9]]
+     *
+     * Slice step 0 at idx 1, results in [1, 2]
+     * Slice step 0, starting at idx 2...length, results in [3]
+     * Merge [3] with step idx 1, [4, 5, 6]
+     *
+     * Example 2:
+     * Target index: 1
+     * Initiator index: 1
+     * Action index: 0
+     * Output: [[1, 2, 3, 4], [5, 6], [7, 8, 9]]
+     *
+     * Example 3:
+     * Target index: 2
+     * Initiator index: 1
+     * Action index: 0
+     * Output: [[1, 2, 3], [4, 5], [6, 7, 8, 9]]
+     */
+    onDropStep: (targetIndex, initiatorIndex, actionIndex) => {
+      if (targetIndex < 0 || targetIndex >= steps.length)
+        throw Error("Cannot drop step index because it does not exist");
+      if (initiatorIndex <= 0 || initiatorIndex >= steps.length)
+        throw Error("Cannot drag specified index because it does not exist");
+
+      const targetStep = steps[targetIndex];
+      const initiatorStep = steps[initiatorIndex];
+
+      // corresponds to example 1 above
+      if (targetIndex < initiatorIndex) {
+        const newTargetStep = targetStep.actions.slice(0, actionIndex + 1);
+        const toMerge = targetStep.actions.slice(
+          actionIndex + 1,
+          targetStep.actions.length
+        );
+        const newInitiator = [...toMerge, ...initiatorStep.actions];
+        setSteps(oldSteps => [
+          ...oldSteps.slice(0, targetIndex),
+          { actions: newTargetStep, name: targetStep.name },
+          { actions: newInitiator, name: initiatorStep.name },
+          ...oldSteps.slice(initiatorIndex + 1, oldSteps.length),
+        ]);
+      }
```

Comment:
IMO it would be worth breaking each case into its own function, just for the sake of readability. We could then also move comments with examples next to each.

Reply:
Check out https://github.com/elastic/synthetics-recorder/pull/175/commits/49837ce84a68f522d35b8640d890473e564db5ad, I moved it to its own directory and broke it up. I was going to move each function to its own file, and still could, but LMK what you think of this so far.